server/service/endpoint_utils.go

```diff
@@ -267,8 +271,54 @@ func (e *UserAuthEndpointer) DELETE(path string, f handlerFunc, v interface{}) {
 	e.handle(path, f, v, "DELETE")
 }
 
-func (e *UserAuthEndpointer) handle(path string, f handlerFunc, v interface{}, verb string) *mux.Route {
-	return e.r.Handle(path, e.makeEndpoint(f, v)).Methods(verb).Name(getNameFromPathAndVerb(verb, path))
+func (e *UserAuthEndpointer) handle(path string, f handlerFunc, v interface{}, verb string) {
+	versions := e.versions
+	if e.startingAtVersion != "" {
+		startIndex := -1
+		for i, version := range versions {
+			if version == e.startingAtVersion {
+				startIndex = i
+				break
+			}
+		}
+		if startIndex == -1 {
+			panic("StartAtVersion is not part of the valid versions")
+		}
+		versions = versions[startIndex:]
+	}
+	if e.endingAtVersion != "" {
+		endIndex := -1
+		for i, version := range versions {
+			if version == e.endingAtVersion {
+				endIndex = i
+				break
+			}
+		}
+		if endIndex == -1 {
+			panic("EndAtVersion is not part of the valid versions")
+		}
+		versions = versions[:endIndex+1]
+	}
+
+	// if a version doesn't have a deprecation version, or the ending version is the latest one, then it's part of the
+	// latest
+	if e.endingAtVersion == "" || e.endingAtVersion == e.versions[len(e.versions)-1] {
+		versions = append(versions, "latest")
+	}
+
+	versionedPath := strings.Replace(path, "/v1/", fmt.Sprintf("/{fleetversion:(?:%s)}/", strings.Join(versions, "|")), 1)
```

Comment:
Should we instead always replace the second segment between `/.../`? I.e. once the v1 version is deprecated, it will feel strange to write the base path like this, it could be a more generic placeholder instead (`/api/_version_/...` or whatever). I don't think it's a big issue, as the usual approach would be to just duplicate an existing path and adjust it for a new feature, but it's just something that might require explanations down the line or to newcomers.

Reply:
Yeah, I agree.